core/internal/featurechecker/features.go

```diff
@@ -0,0 +1,115 @@
+package featurechecker
+
+import (
+	"context"
+	"sync"
+
+	"github.com/Khan/genqlient/graphql"
+	"github.com/wandb/wandb/core/internal/gql"
+	"github.com/wandb/wandb/core/internal/observability"
+	spb "github.com/wandb/wandb/core/pkg/service_go_proto"
+)
+
+// ServerFeatures is responsible for providing the capabilities of a server
+// Features and capabilities are retrieved from the server via a GraphQL query
+type ServerFeatures struct {
+	Features map[spb.ServerFeature]ServerFeature
+
+	ctx           context.Context
+	graphqlClient graphql.Client
+	mu            sync.Mutex
+}
+
+type ServerFeature struct {
+	DefaultValue string
+	Description  string
+	Enabled      bool
+	Name         string
+}
+
+func NewServerFeatures(
+	ctx context.Context,
+	graphqlClient graphql.Client,
+) *ServerFeatures {
+	var features map[spb.ServerFeature]ServerFeature
+
+	// If graphqlClient is nil, we won't ever be able to query the server
+	// So create an empty map which will default to false for all features when a key is not found
+	if graphqlClient == nil {
+		features = make(map[spb.ServerFeature]ServerFeature)
+	}
+
+	return &ServerFeatures{
+		Features:      features,
+		ctx:           ctx,
+		graphqlClient: graphqlClient,
+		mu:            sync.Mutex{},
+	}
+}
+
+func (sf *ServerFeatures) getServerFeatures() {
```

Comment:
how about using sync.Once and calling the graphql and handling logic there that way you only do it once and you can populate the results, you can also store the error if you want to return it

Reply:
```
func defaultFetchFn(
	ctx context.Context,
	client graphql.Client,
) (*ServerFeatures, error) {
	// Query the server for the features provided by the server
	resp, err := gql.ServerFeaturesQuery(
		ctx,
		client,
		gql.RampIDTypeUsername,
	)
	if err != nil {
		return nil, err
	}

	features := make(ServerFeatures)

	for _, f := range resp.ServerInfo.FeatureFlags {
		featureName := spb.ServerFeature(spb.ServerFeature_value[f.RampKey])
		features[featureName] = Feature{
			Name:    f.RampKey,
			Enabled: f.IsEnabled,
		}
	}

	return &features, nil
}

func (fc *FeaturesCach) fetchFeatures() (*ServerFeatures, error) {
	fc.once.Do(func() {
		features, err := defaultFetchFn(fc.ctx, fc.client)
		fc.mu.Lock()
		defer fc.mu.Unlock()
		fc.features = features
		fc.err = err
	})
	fc.mu.RLock()
	defer fc.mu.RUnlock()
	return fc.features, fc.err
}
```

something like this

Reply:
After reading more about `sync.Once` looks like we dont even need a mutex here right?

Reply:
ah this is just a sketch, feel free to change it, but i think you would still need some locks for accessing the features.

Reply:
Hmm I dont think we do, but I can add for safety?

During loading of features we will synchronize all concurrent operations from writing/reading data as sync.once.Do will prevent multiple operations on our `features` cache. 
After updating we should not have any more updates to the features cache, so reading the data should idempotent.